### ISSUE TITLE
feature request: export / import contacts and correspondence.

### ISSUE THREAD

**Issue by NektoNektovich** [evilcorpltd/aTox#1047]
feature request: export / import contacts and correspondence.

**Comment by giiutfff** [comment]
I second this.
The current tox export/import doesn't include history and files. which means when switching to a new device, they can't be transferred to new device.
A simple option to archive/restore the whole app data (as a zip file) would be practical and easy to implement.
Kind regards

**Comment by robinlinden** [comment]
Yeah, I think that's a reasonable thing to add. @theAkito has added the ability to export your messages (per contact), so I guess we'll continue along that route and add support for importing the same format in the future. :)

**Comment by NektoNektovich** [comment]
Thanks! but about the received and transferred files, I would still really
like to have the setting for transferring it to a memory card like telegram
along the path /storage/memory-card/Android/data/ltd.evilcorp.atox/cache
🤓☺️

суббота, 5 ноября 2022 г. пользователь Robin Lindén <
***@***.***> написал:
> Yeah, I think that's a reasonable thing to add. @theAkito has added the
ability to export your messages (per contact), so I guess we'll continue
along that route and add support for importing the same format in the
future. :)
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.<
https://ci5.googleusercontent.com/proxy/kCEITNnLFGRvox1PaGE-BXE0Y92pxECLD6ywvifPBum4sEO9l91_1d_6FOM7ugMz2VAyIMy-Nrtjw1j61Hc_ZEJ_LdQapOmOnZ8DNFRNrDEOOp0e6CpEgQfZIvV7drWwdLnp3-4lifDiPV-J6BuV0m0GLG55c6gb_HzoiL0Gwzz3Q44csLU4EkLPer-2fN700w-KUwvnrzgcSf-WWbSoKvbroYRG-3dvvLngmRxolw=s0-d-e1-ft#https://github.com/notifications/beacon/AMZDE6CPEXTWXJCULY6VWN3WGWDLNA5CNFSM5XOKK6FKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOJW7BFDQ.gif>Message
ID: ***@***.***>

**Comment by giiutfff** [comment]
> Yeah, I think that's a reasonable thing to add. @theAkito has added the ability to export your messages (per contact), so I guess we'll continue along that route and add support for importing the same format in the future. :)

I believe it will be less convenient to import all the data, if one has a long contact list. Although export messages for particular contact(s) could be useful. A method to backup/restore the whole app's data(that also includes transfered files and settings) would be nice.




btw, that json format is not really human readable, and super redundant. 

>   "timestamp" : "2022-10-30T02-32-04",

What time zone is this?

> "timestamp" : 1667079807880

And mixed with different time format .

> "publicKey" : "4B9CD365CFF0F4BF0A13ACFE71C612AECF26A4699F04D1384AEA8F5A52734106",

Since there is only one contact, no need to include this information for every massage.


Kind regards

**Comment by theAkito** [comment]
> btw, that json format is not really human readable, and super redundant. 
> 
> >   "timestamp" : "2022-10-30T02-32-04",
> 
> What time zone is this?
> 
> > "timestamp" : 1667079807880
> 
> And mixed with different time format .
> 
> > "publicKey" : "4B9CD365CFF0F4BF0A13ACFE71C612AECF26A4699F04D1384AEA8F5A52734106",
> 
> Since there is only one contact, no need to include this information for every massage.

I was aware of that when I implemented it and thought about making it less redundant. However, since I am planning to make these backups importable at some point, I didn't want to change the message format, as long as I don't have a clear plan on how to do the import, so I just kept the messages verbatim, just as they are actually in the app.
Each message object is currently a 1:1 representation of what is actually, in reality, within aTox.

That said, I must emphasize that the Import/Export functionality is obviously overall still a work in progress. Especially, as long as the Import functionality is not yet available, you cannot judge the format as if it would stay the same forever from now on. It will change. Hence there is also a backup version, so the app can import & export older backups.

As for the human readability, I'm not sure if you specifically applied that to the redunandcy or the overall serialization method.

As explained above, the redundancy will be reduced in the future, but the format, i. e. JSON, won't change. It is human readable, when pretty printed, and one of the best methods available.

**Comment by giiutfff** [comment]
Thanks for your explanation. A human readable format would be something like a HTML document with visual decoration like bubbles in my humble opinion, although it may not suitable for import back. And unix timestamp is absolutely not human readable under any circumstance. 
It is understandable that the data stored in app include much more information. However the export function is only use in two ways as I can imagine: 1. for reading the data outside the app; 2. for backup (import back to the app)
For purpose 1, a human readable format is favoured. For 2, a 1:1 copy should do. As your implementation is done as per contact export, so I assume it's for purpose 1. That's why I gabbled about readability.

And one more thing, xml+xsl might be another best method available.
Kind regards
